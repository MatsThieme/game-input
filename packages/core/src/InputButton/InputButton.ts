export class InputButton {
    private _isSetDown: boolean | undefined;
    private _down: boolean;
    private _wasDown: boolean;

    /**
     * Store state information of a button.
     */
    public constructor() {
        this._isSetDown = false;
        this._down = false;
        this._wasDown = false;
    }

    /**
     * Returns whether the button was down in the last frame.
     */
    public get down(): boolean {
        return this._down;
    }

    /**
     * Returns whether the button was down in the second last frame.
     */
    public get wasDown(): boolean {
        return this._wasDown;
    }

    /**
     * Returns whether the button was down in the last frame and was not down in the second last frame.
     */
    public get click(): boolean {
        return this._down && !this._wasDown;
    }

    /**
     * Returns whether the button was down in the last and second last frame.
     */
    public get clicked(): boolean {
        return this._down && this._wasDown;
    }

    /**
     * @internal
     */
    public setDown(down: boolean): void {
        this._isSetDown = down;
    }

    /**
     * @internal
     */
    public update(): void {
        this._wasDown = this._down;

        if (this._isSetDown !== undefined) this._down = this._isSetDown;

        this._isSetDown = undefined;
    }
}
