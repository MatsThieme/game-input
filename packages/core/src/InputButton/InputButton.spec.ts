import { InputButton } from "./InputButton";
import { describe, it, expect } from "vitest";

describe("InputButton", () => {
    it("should update correctly", () => {
        const button = new InputButton();

        snapBtn(button);

        button.update();

        expect(snapBtn(button)).toMatchSnapshot();

        button.setDown(true);

        button.update();

        expect(snapBtn(button)).toMatchSnapshot();

        button.update();

        expect(snapBtn(button)).toMatchSnapshot();

        button.setDown(false);

        button.update();

        expect(snapBtn(button)).toMatchSnapshot();

        button.update();

        expect(snapBtn(button)).toMatchSnapshot();
    });
});

function snapBtn(button: InputButton): void {
    expect({
        down: button.down,
        wasDown: button.wasDown,
        click: button.click,
        clicked: button.clicked,
    }).toMatchSnapshot();
}
