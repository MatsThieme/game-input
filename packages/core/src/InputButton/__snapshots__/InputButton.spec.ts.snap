// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputButton > should update correctly 1`] = `
{
  "click": false,
  "clicked": false,
  "down": false,
  "wasDown": false,
}
`;

exports[`InputButton > should update correctly 2`] = `
{
  "click": false,
  "clicked": false,
  "down": false,
  "wasDown": false,
}
`;

exports[`InputButton > should update correctly 3`] = `undefined`;

exports[`InputButton > should update correctly 4`] = `
{
  "click": true,
  "clicked": false,
  "down": true,
  "wasDown": false,
}
`;

exports[`InputButton > should update correctly 5`] = `undefined`;

exports[`InputButton > should update correctly 6`] = `
{
  "click": false,
  "clicked": true,
  "down": true,
  "wasDown": true,
}
`;

exports[`InputButton > should update correctly 7`] = `undefined`;

exports[`InputButton > should update correctly 8`] = `
{
  "click": false,
  "clicked": false,
  "down": false,
  "wasDown": true,
}
`;

exports[`InputButton > should update correctly 9`] = `undefined`;

exports[`InputButton > should update correctly 10`] = `
{
  "click": false,
  "clicked": false,
  "down": false,
  "wasDown": false,
}
`;

exports[`InputButton > should update correctly 11`] = `undefined`;
